File: modules/commercial_builder/routes/vox.py
"""The Vox explainer step — a beat list instead of a storyboard.

Every other commercial type in this module is a sequence of scenes each
holding a piece of footage, assembled by Creatomate. A Vox explainer is not:
it is an editorial collage rendered **whole** by `hub/hyperframes.py` from a
beat list, so it takes this one step in place of Concepts, Blueprint, Voice
and CTA.

Four rules here, each of them a way this goes quietly wrong.

**Nothing is rendered by arriving.** Generating the beats is one press and
rendering is another, because a render is minutes of headless Chrome and the
beat list is the only thing anybody can correct before it. Same "approve
before spend" shape every expensive step in this Hub uses.

**An edit is validated exactly as a generation is.** `vox_spec.validate()`
runs over what a person typed as well as over what the model wrote — a rule
the form keeps while the write breaks it is not a rule, and this is the join
where a typed `seconds` of "about 8" reaches a template that will read it as
zero.

**Every write to `project.script` merges.** Assigning a fresh dict over it is
the `set_music` trap this module has already paid for once, where a
voiceover URL written by one panel was wiped by the next save on another.

**The beats live in `script_json`, never in a new column.** `create_all()`
creates missing tables and never adds a column to an existing one, so a
`beats` column would exist on every local SQLite run and be silently absent
on the live Postgres — the trap `hub/client_key.py` states at length.
"""

from flask import Blueprint, jsonify, request

from .. import vox_spec
from ..db import db
from ..models import Client, CommercialProject
from ..services import openai_service
from hub import hyperframes

bp = Blueprint("cb_vox", __name__, url_prefix="/api")


# Writes on this blueprint that deliberately record nothing, with the reason.
# The line this module records on is whether something reaches the client's
# own Cloudinary tree or changes their record; a beat list is a draft of a
# piece nobody has rendered yet, and `submit_render` records the render.
HOUSEKEEPING_ROUTES = {
    "generate_vox_beats": "drafts a beat list for somebody to read. Billed as "
                          "an OpenAI call, which `hub/ai.py` records, and "
                          "nothing has been made for the client until it is "
                          "rendered.",
    "save_vox_beats": "edits that draft.",
}


def _is_vox(project):
    return (project.commercial_type or "") == vox_spec.COMMERCIAL_TYPE


def _not_vox():
    return jsonify({"ok": False, "error": (
        "This spot is not a Vox explainer, so it has no beat list — it is "
        "built as a storyboard of scenes.")}), 400


def _state(project):
    """Everything a screen needs about this project's beats, in one shape.

    Composed here rather than in the template so the page cannot come to
    disagree with the route about whether a list is renderable — the same
    reason `preview.js` was made to read the server's own severity rather
    than keeping a set of its own.
    """
    script = project.script or {}
    beats = script.get("beats") or []
    seconds = vox_spec.beats_seconds(beats)
    return {
        "beats": beats,
        "source": script.get("beats_source") or "",
        "seconds": seconds,
        "duration": hyperframes.vox_duration_verdict(seconds if beats else None),
        "enough_beats": len(beats) >= vox_spec.MIN_BEATS,
        "min_beats": vox_spec.MIN_BEATS, "max_beats": vox_spec.MAX_BEATS,
        "treatments": vox_spec.TREATMENTS,
        "source_kinds": vox_spec.SOURCE_KINDS,
        "render_service": hyperframes.check(),
    }


@bp.get("/projects/<int:project_id>/vox")
def read_vox(project_id):
    project = CommercialProject.query.get_or_404(project_id)
    if not _is_vox(project):
        return _not_vox()
    return jsonify({"ok": True, **_state(project)})


@bp.post("/projects/<int:project_id>/vox/beats")
def generate_vox_beats(project_id):
    """Draft the beat list from a topic, a document or a link.

    The one step in this feature that needs a model. What comes back is
    validated against the template's contract before it is stored, because
    the model writes JSON and the template consumes JSON and nothing between
    them otherwise checks that the two agree.
    """
    project = CommercialProject.query.get_or_404(project_id)
    if not _is_vox(project):
        return _not_vox()
    client = Client.query.get_or_404(project.client_id)
    data = request.get_json(force=True) or {}

    kind = vox_spec.clean_source_kind(data.get("source_kind"))
    text = str(data.get("source_text") or "").strip()
    link = str(data.get("link") or "").strip()

    result = openai_service.generate_vox_beats(
        kind, text, client.to_dict(),
        title=project.title or client.name,
        total_seconds=project.length_seconds,
        link=link)

    if not result.get("beats"):
        # Nothing usable came back and nothing is stored. A beat list a
        # person cannot read is not a draft, and overwriting a good one with
        # it would lose work to a bad answer from a provider.
        return jsonify({"ok": False, "error": result.get("error")
                        or "Nothing could be built into an explainer from that.",
                        "dropped": result.get("dropped") or []}), 502

    # Merged, never assigned. The `set_music` trap: a fresh dict over
    # `script` would destroy anything else written there.
    script = dict(project.script or {})
    script["beats"] = result["beats"]
    script["beats_source"] = result.get("source") or ""
    script["vox_source"] = {"kind": kind, "link": link, "text": text[:4000]}
    project.script = script
    if project.status == "draft":
        project.status = "scripted"
    db.session.commit()

    return jsonify({"ok": True, "dropped": result.get("dropped") or [],
                    "note": result.get("error") or "", **_state(project)})


@bp.put("/projects/<int:project_id>/vox/beats")
def save_vox_beats(project_id):
    """Keep what somebody edited, held to the same contract.

    Validated rather than trusted for the same reason the generated list is:
    a `seconds` typed as text and a treatment nobody offered both render, and
    neither errors.
    """
    project = CommercialProject.query.get_or_404(project_id)
    if not _is_vox(project):
        return _not_vox()
    data = request.get_json(force=True) or {}

    checked = vox_spec.validate(data.get("beats"),
                                total_seconds=project.length_seconds)
    if not checked["beats"]:
        return jsonify({"ok": False, "error": (
            "That leaves no beats at all. An explainer needs at least "
            f"{vox_spec.MIN_BEATS}."), "dropped": checked["dropped"]}), 400

    script = dict(project.script or {})
    script["beats"] = checked["beats"]
    # An edited list is somebody's, whoever drafted it. Left reading "ai" it
    # would go on crediting the model for a beat a person wrote.
    script["beats_source"] = "edited"
    project.script = script
    db.session.commit()
    return jsonify({"ok": True, "dropped": checked["dropped"], **_state(project)})

File: modules/commercial_builder/static/js/vox.js
/* The Vox explainer's beat list.

   Two rules shape this file. Nothing renders from here — writing the beats
   and rendering them are separate presses, because a render is minutes of
   headless Chrome and this list is the only thing anybody can correct before
   it. And the server is what decides whether a list is usable: the duration
   verdict, the treatments and the beat bounds all come back on every
   response rather than being recomputed here, or the page and the route come
   to disagree about whether a spot can be rendered. */
(() => {
  const root = document.getElementById("vox-root");
  if (!root) return;
  const projectId = root.dataset.projectId;

  function data(id) {
    const node = document.getElementById(id);
    if (!node) return [];
    try { return JSON.parse(node.textContent) || []; } catch (e) { return []; }
  }
  const TREATMENTS = data("vox-treatments-data");
  const KINDS = data("vox-kinds-data");

  const kindSel = document.getElementById("vox-kind");
  const kindHint = document.getElementById("vox-kind-hint");
  const textField = document.getElementById("vox-text-field");
  const linkField = document.getElementById("vox-link-field");
  const list = document.getElementById("vox-beats");
  const durationNote = document.getElementById("vox-duration");

  /* The model's own beats, held here between edits. Read back out of the
     inputs before any redraw -- a container of live inputs rebuilt with
     innerHTML loses whatever was half-typed in it, which is the trap the SEO
     client page's alt-text and FAQ editors each had. */
  let beats = [];

  function syncKind() {
    const kind = kindSel.value;
    const match = KINDS.find((k) => k.id === kind);
    kindHint.textContent = match ? match.hint : "";
    linkField.style.display = kind === "link" ? "" : "none";
    textField.style.display = kind === "link" ? "none" : "";
  }
  kindSel.addEventListener("change", syncKind);
  syncKind();

  function harvest() {
    /* Read the open editors back into the model before anything redraws.
       Only fields somebody actually typed in, which is why each carries the
       value it was drawn with: comparing against the model would treat a
       whole freshly-written list as edited. */
    list.querySelectorAll("[data-beat-index]").forEach((row) => {
      const i = parseInt(row.dataset.beatIndex, 10);
      if (!beats[i]) return;
      const head = row.querySelector(".beat-headline");
      const sup = row.querySelector(".beat-support");
      const tre = row.querySelector(".beat-treatment");
      const sec = row.querySelector(".beat-seconds");
      const src = row.querySelector(".beat-source");
      if (head) beats[i].headline = head.value;
      if (sup) beats[i].support = sup.value;
      if (tre) beats[i].treatment = tre.value;
      if (sec) beats[i].seconds = parseFloat(sec.value) || 0;
      if (src) beats[i].source = src.value;
    });
  }

  function render() {
    list.innerHTML = "";
    if (!beats.length) {
      list.appendChild(CB.el('<div class="cb-empty">No beats yet.</div>'));
      return;
    }
    beats.forEach((b, i) => {
      const row = CB.el(`<div class="cb-card" data-beat-index="${i}"
        style="margin-bottom:8px;padding:10px;">
        <div class="cb-flex-between" style="margin-bottom:6px;">
          <strong>Beat ${i + 1}</strong>
          <button class="cb-btn cb-btn-sm cb-btn-danger beat-delete">Remove</button>
        </div>
        <div class="cb-field"><label class="cb-label">Headline</label>
          <input type="text" class="beat-headline"></div>
        <div class="cb-field"><label class="cb-label">Support</label>
          <input type="text" class="beat-support"></div>
        <div style="display:flex;gap:8px;flex-wrap:wrap;">
          <div class="cb-field"><label class="cb-label">Treatment</label>
            <select class="beat-treatment"></select></div>
          <div class="cb-field"><label class="cb-label">Seconds</label>
            <input type="number" step="0.5" min="0" class="beat-seconds" style="width:90px;"></div>
          <div class="cb-field" style="flex:1;"><label class="cb-label">Source</label>
            <input type="text" class="beat-source" placeholder="Required for a quote"></div>
        </div>
      </div>`);
      row.querySelector(".beat-headline").value = b.headline || "";
      row.querySelector(".beat-support").value = b.support || "";
      row.querySelector(".beat-seconds").value = b.seconds || 0;
      row.querySelector(".beat-source").value = b.source || "";
      const sel = row.querySelector(".beat-treatment");
      TREATMENTS.forEach((t) => {
        const opt = document.createElement("option");
        opt.value = t.id;
        opt.textContent = t.label;
        opt.title = t.hint;
        sel.appendChild(opt);
      });
      sel.value = b.treatment || (TREATMENTS[0] || {}).id || "";
      row.querySelector(".beat-delete").addEventListener("click", () => {
        harvest();
        beats.splice(i, 1);
        render();
      });
      list.appendChild(row);
    });
  }

  /* The verdict is the server's. Recomputing the window here would be a
     second answer to "can this be rendered", and the two disagree the day
     either is edited. */
  function paintState(state) {
    if (!state) return;
    beats = state.beats || [];
    const d = state.duration || {};
    durationNote.textContent = d.message || "";
    render();
  }

  function dropNote(target, dropped, note) {
    const box = document.getElementById(target);
    box.innerHTML = "";
    if (note) box.appendChild(CB.el(`<p class="cb-hint">${CB.escapeHtml(note)}</p>`));
    /* What was thrown away and why. A list that quietly gets shorter is an
       explainer missing exactly the beat somebody wanted made. */
    (dropped || []).forEach((d) => {
      box.appendChild(CB.el(
        `<p class="cb-hint warn">Dropped: ${CB.escapeHtml(d.reason || "")}</p>`));
    });
  }

  document.getElementById("vox-generate").addEventListener("click", async (e) => {
    /* CB.working() into the note box rather than S1Think.attach() onto the
       button: attach() APPENDS its status box into whatever it is given, so
       on a <button> the mark lands inside the label. The module's own panel
       is the shape every other wait in this tool draws. */
    const btn = e.target;
    const note = document.getElementById("vox-generate-note");
    note.innerHTML = CB.working("concepts", "Writing the beats\u2026",
      "The material is read and turned into an outline. Nothing renders yet \u2014 "
      + "you read the beats and press Render on the next step.");
    btn.disabled = true;
    try {
      const body = {
        source_kind: kindSel.value,
        source_text: document.getElementById("vox-text").value,
        link: document.getElementById("vox-link").value,
      };
      const state = await CB.api(`/api/projects/${projectId}/vox/beats`,
                                 { method: "POST", body });
      paintState(state);
      dropNote("vox-generate-note", state.dropped, state.note);
    } catch (err) {
      note.innerHTML = "";      // CB.api has already surfaced the reason
    } finally {
      btn.disabled = false;
    }
  });

  document.getElementById("vox-add").addEventListener("click", () => {
    harvest();
    beats.push({ headline: "", support: "", treatment: (TREATMENTS[0] || {}).id || "",
                 seconds: 0, source: "", image_query: "" });
    render();
  });

  document.getElementById("vox-save").addEventListener("click", async (e) => {
    harvest();
    const btn = e.target;
    btn.disabled = true;
    try {
      const state = await CB.api(`/api/projects/${projectId}/vox/beats`,
                                 { method: "PUT", body: { beats } });
      paintState(state);
      dropNote("vox-save-note", state.dropped, "");
      CB.toast("Beats saved.");
    } catch (err) {
      /* CB.api has already surfaced the reason. */
    } finally {
      btn.disabled = false;
    }
  });

  (async function load() {
    try {
      paintState(await CB.api(`/api/projects/${projectId}/vox`));
    } catch (e) { /* the empty state already says there is nothing */ }
  })();
})();
